src/lib.rs

```diff
@@ -532,21 +563,39 @@ impl ConfiguredOidc {
         let original_path = self.get_http_request_header(":path").unwrap_or_default();
         let original_path_encoded = base64engine.encode(original_path.as_bytes());
 
-        debug!("No cookie found or invalid, redirecting to authorization endpoint.");
+        debug!("no cookie found or invalid, redirecting to authorization endpoint");
 
         // Generate PKCE code verifier and challenge
         let pkce_verifier = pkce::code_verifier(128);
         let pkce_verifier_string = String::from_utf8(pkce_verifier.clone()).unwrap();
         let pkce_encoded = base64engine.encode(pkce_verifier_string.as_bytes());
         let pkce_challenge = pkce::code_challenge(&pkce_verifier);
 
+        // Generate state
+        let state_string = String::from_utf8(pkce::code_verifier(128)).unwrap();
+        debug!("state: {}", state_string);
+
+        // Store request id in shared data
+        let request_id = self.get_http_request_header("x-request-id").unwrap_or_default();
+        match self.set_shared_data(
+        format!("state-{}", request_id).as_str(),
+        Some(state_string.as_bytes()), None) {
+            Ok(_) => {
+                debug!("state stored in shared data");
+            }
+            Err(e) => {
+                warn!("storing state in shared data failed: {:?}", e);
+            }
+        }
```

Comment:
I think this is problematic for two reasons:
* The shared data is never removed, so this is a memory leak
* It's not stateless. If for some reason we are running multiple instances of the proxy behind a load balancer, or if the proxy has to restart in the middle of a flow, then this will fail

I *think* it is sufficiently secure to store the state in a cookie instead. This is just a mitigation against PKCE downgrade anyway. If an attacker wants to inject their own auth-code, they would then need to steal the cookie as well to have the state. But if they can steal the state cookie, they can also steal the PKCE verifier cookie. To protect against that scenario, these cookies would also have to be encrypted. But if the attacker can steal your cookies, you probably have bigger problems already.

Reply:
Hmm, so are you saying the state should be stored client side in an encrypted cookie? And not in the Envoy Container? Sounds ok to me. However, we should discuss if this is needed at all or if SATOSA is the problem here, because (as you said), SATOSA should enforce PKCE which makes the nonce "useless".

Reply:
Hmm, so are you saying the state should be stored client side in an encrypted cookie? And not in the Envoy Container? Sounds ok to me. However, we should discuss if this is needed at all or if this is the responsibility of OIDC providers to enforce PKCE.